Use observers for callbacks (#249)
TheBevyFlock/bevy_quickstart: docs/design.md

### Pattern

When spawning an entity that can be interacted with, such as a button that can be pressed,
use an observer to handle the interaction:

```rust
fn spawn_button(mut commands: Commands) {
    // See the Widgets pattern for information on the `button` method
    commands.button("Pay up!").observe(pay_money);
}

> fn pay_money(_trigger: Trigger<OnPress>, /* Bevy query parameters */) {

Reviewer: ```suggestion
fn pay_money(_trigger: Trigger<OnPress>, /* Bevy system parameters */) {
```
Although this would be more clear with e.g. a dummy `mut money: ResMut<Money>` and then `money.0 += 100.0;` in the body.

Author: [`6d794b2` (#249)](https://github.com/TheBevyFlock/bevy_quickstart/pull/249/commits/6d794b242347929155499c766cf7846d1aceef8e)

    // Handle the interaction
}
```

The event `OnPress`, which is [defined in this template](../src/theme/interaction.rs),

> is triggered when the button's [`Interaction`](https://docs.rs/bevy/latest/bevy/prelude/enum.Interaction.html)
> component becomes [`Interaction::Pressed`](https://docs.rs/bevy/latest/bevy/prelude/enum.Interaction.html#variant.Pressed).

Reviewer: ```suggestion
is triggered when the button is [`Interaction::Pressed`](https://docs.rs/bevy/latest/bevy/prelude/enum.Interaction.html#variant.Pressed).
```
Optional.

### Reasoning

This pattern is inspired by [bevy_mod_picking](https://github.com/aevyrie/bevy_mod_picking).

> By coupling the system handling the interaction to the entity as an observer,

Reviewer: ```suggestion
By pairing the system handling the interaction to the entity as an observer,
```
Optional. "coupling" reminds me of "code coupling" which is an anti-pattern.

Author: Wouldn't it be "pairing A with B" instead of "pairing A to B"?

Author: [`6fd702b` (#249)](https://github.com/TheBevyFlock/bevy_quickstart/pull/249/commits/6fd702b3fda95461e7653a5997d1ea6634f660ab)

Reviewer: I think either way is correct.

the code running on interactions can be scoped to the exact context of the interaction.

For example, the code for what happens when you press a *specific* button is directly attached to that exact button.

This also keeps the interaction logic close to the entity that is interacted with,
allowing for better code organization. If you want multiple buttons to do the same thing,

> consider triggering a secondary event in their callbacks.

Reviewer: You can just use the same observer, no? `button1.observe(foo); button2.observe(foo);` where `foo` takes `Trigger<OnPress>`.

Reviewer 2: I had the same thought, decided to not comment as only "consideration" is suggested 😅

Author: [`0f27c74` (#249)](https://github.com/TheBevyFlock/bevy_quickstart/pull/249/commits/0f27c74a8c4d91d7c7c469195a32453b6e449ee2)

### Pattern

When spawning an entity that can be interacted with, such as a button that can be pressed,
use an observer to handle the interaction:

```rust
fn spawn_button(mut commands: Commands) {
    // See the Widgets pattern for information on the `button` method
    commands.button("Pay up!").observe(pay_money);
}

fn pay_money(_trigger: Trigger<OnPress>, /* Bevy query parameters */) {
    // Handle the interaction
}
```

The event `OnPress`, which is [defined in this template](../src/theme/interaction.rs),
is triggered when the button is [`Interaction::Pressed`](https://docs.rs/bevy/latest/bevy/prelude/enum.Interaction.html#variant.Pressed).

> If your interactions mostly consist of changing the game state, consider using the following helper function:

Reviewer: ```suggestion
If you have many interactions that only change a state, consider using the following helper function:
```
Reasoning: "game state" is redundant, every state is a game state; and the interactions have to be _only_ changing the state, not "mostly consisting of" changing the state, for the helper function to be useful.

```rust
fn spawn_button(mut commands: Commands) {
    commands.button("Play the game").observe(change_state(Screen::Playing));
}

> fn change_state<S: FreelyMutableState>(

Reviewer: Nit: `enter_state` reads better IMO.

Author: [`724986e` (#249)](https://github.com/TheBevyFlock/bevy_quickstart/pull/249/commits/724986e4c60592832a20006861ab29b5440193dc)
